internal/expr/query.go

```diff
@@ -36,93 +36,142 @@ func (pe *PreparedExpr) Query(args ...any) (ce *QueryExpr, err error) {
-		if !ok {
-			return nil, fmt.Errorf(`type %s not found, have: %s`, in.structType.Name(), strings.Join(typeNames, ", "))
+	for i, typeElement := range pe.inputs {
+		switch te := typeElement.(type) {
+		case field:
+			v, ok := typeValue[te.structType]
+			if !ok {
+				return nil, fmt.Errorf(`type %s not found, have: %s`, te.structType.Name(), strings.Join(typeNames, ", "))
+			}
+			qargs = append(qargs, sql.Named("sqlair_"+strconv.Itoa(i), v.FieldByIndex(te.index).Interface()))
+		case mapKey:
+			v, ok := m[te.name]
+			if !ok {
+				return nil, fmt.Errorf(`map does not contain key %q`, te.name)
+			}
+			qargs = append(qargs, sql.Named("sqlair_"+strconv.Itoa(i), v))
+		default:
+			return nil, fmt.Errorf("internal error: field type %T not supported", te)
 		}
-		named := sql.Named("sqlair_"+strconv.Itoa(i), v.FieldByIndex(in.index).Interface())
-		qargs = append(qargs, named)
 	}
-
 	return &QueryExpr{outputs: pe.outputs, sql: pe.sql, args: qargs}, nil
 }
 
+type MapDecodeInfo struct {
+	mapPtrs  []any
+	keyIndex map[string]int
+	m        map[string]any
+}
+
 // ScanArgs returns list of pointers to the struct fields that are listed in qe.outputs.
 // All the structs mentioned in the query must be in outputArgs.
 // All outputArgs must be structs.
-func (qe *QueryExpr) ScanArgs(columns []string, outputArgs []any) ([]any, error) {
+func (qe *QueryExpr) ScanArgs(columns []string, outputArgs []any) ([]any, *MapDecodeInfo, error) {
+	var inQuery = make(map[reflect.Type]bool)
+	for _, typeElement := range qe.outputs {
+		switch te := typeElement.(type) {
+		case field:
+			inQuery[te.structType] = true
+		}
+	}
+
+	var typeDest = make(map[reflect.Type]reflect.Value)
+	var m map[string]any
 	outputVals := []reflect.Value{}
 	for _, outputArg := range outputArgs {
 		if outputArg == nil {
-			return nil, fmt.Errorf("need pointer to struct, got nil")
+			return nil, nil, fmt.Errorf("need map or pointer to struct, got nil")
 		}
 		outputVal := reflect.ValueOf(outputArg)
 		if outputVal.Kind() != reflect.Pointer {
-			return nil, fmt.Errorf("need pointer to struct, got %s", outputVal.Kind())
+			return nil, nil, fmt.Errorf("need map or pointer to struct, got %s", outputVal.Kind())
 		}
 		if outputVal.IsNil() {
-			return nil, fmt.Errorf("got nil pointer")
+			return nil, nil, fmt.Errorf("got nil pointer")
 		}
 		outputVal = reflect.Indirect(outputVal)
-		if outputVal.Kind() != reflect.Struct {
-			return nil, fmt.Errorf("need pointer to struct, got pointer to %s", outputVal.Kind())
+		switch k := outputVal.Kind(); k {
+		case reflect.Struct:
+			if !inQuery[outputVal.Type()] {
+				return nil, nil, fmt.Errorf("type %q does not appear as an output type in the query", outputVal.Type().Name())
+			}
+			typeDest[outputVal.Type()] = outputVal
+		case reflect.Map:
+			if m != nil {
+				return nil, nil, fmt.Errorf(`found multiple map types`)
+			}
+			switch arg := outputArg.(type) {
+			case map[string]any:
+				m = arg
+			case *map[string]any:
```

Comment:
If only one of `map[string]any` and `*map[string]any` were supported (as above), this would also need changing